### ISSUE TITLE
Type inference does not infer all properties of a record via multiple member expressions

### ISSUE THREAD

**Issue by nathanielc** [influxdata/flux#4051]
This query causes this panic:

```
from(bucket: "bucket")
    |> range(start: -10m)
    |> filter(fn: (r) => r._measurement == "m")
    |> map(
        fn: (r) => {
            a = r.a
            b = r.b
            c = r._value + 1
            return r
        },
    )
```

Note that `r.a` and `r.b` do not exist on the data. The `r._value` column does exist and it is a float.

Stacktrace and panic

```
panic: unexpected kind: got "float" expected "int", 
trace: goroutine 5107 [running]: runtime/debug.Stack() 
/usr/local/go/src/runtime/debug/stack.go:24 +0x65 github.com/influxdata/flux/values.unexpectedKind.As({0x0, 0x5}, {0x370eb60, 0xc0049bb538})
/idpe/vendor/github.com/influxdata/flux/values/values.go:347 +0x65 github.com/influxdata/flux/values.unexpectedKind.Error({0xc02e148c30, 0x0})
/idpe/vendor/github.com/influxdata/flux/values/values.go:341 +0x2e github.com/influxdata/idpe/pkg/influxerror.(*Error).Error(0xc0053c8840)
/idpe/pkg/influxerror/errors.go:111 +0x19f github.com/influxdata/idpe/query/queue.(*Handler).handleError(0xc004740360, {0x3c1e0b8, 0xc0050d2660}, {0x3bd3880, 0xc0053c8840}, {0x3c1fcb8, 0xc0001211e0})
/idpe/query/queue/worker.go:179 +0x354 github.com/influxdata/idpe/query/queue.(*Handler).Handle(0xc004740360, {0x3c1e0b8, 0xc00510b050}, {0x3c1fcb8, 0xc0001211e0}, 0xc00510c640)
/idpe/query/queue/worker.go:150 +0x9bb github.com/influxdata/idpe/pkg/queue/worker.(*Dispatcher).handle(0xc0001415e0, {0x3c1e0b8, 0xc00510b050}, 0xc0001211e0, 0x12)
/idpe/pkg/queue/worker/redis_dispatcher.go:382 +0x282 github.com/influxdata/idpe/pkg/queue/worker.(*Dispatcher).runOnce(0xc0001415e0, {0x3c1e010, 0xc0019f5780})
/idpe/pkg/queue/worker/redis_dispatcher.go:353 +0xf96 github.com/influxdata/idpe/pkg/queue/worker.(*Dispatcher).stateRun(0xc0001415e0, {0x3c1e010, 0xc0019f5780})
/idpe/pkg/queue/worker/redis_dispatcher.go:238 +0x5d github.com/influxdata/idpe/pkg/queue/worker.(*Dispatcher).Run(0xc0001415e0, {0x3c1e010, 0xc0019f5780})
/idpe/pkg/queue/worker/redis_dispatcher.go:398 +0x186 github.com/influxdata/idpe/internal/cmd/queryd.runQueryd.func4()
/idpe/internal/cmd/queryd/queryd.go:821 +0x2e created by github.com/influxdata/idpe/internal/cmd/queryd.runQueryd
/idpe/internal/cmd/queryd/queryd.go:820 +0x3912   
```

I experimented with this query and almost any change to this query causes the panic to go away. 

For example removing any assignment statement, or changing the `r.a` and `r.b` to be the same column both cause the panic to go away. My guess is we have some kind of collision where type vars are getting mixed up. Perhaps in the flatbuffer serialization or somewhere in type inference.

**Comment by nathanielc** [comment]
I ran the above query through type inference and confirmed that it infers the type of `r` to be a record with this type:

```
{A with a:B, b:C, _value:int}
```

of note is that it infers `_value` to be an int. This is expected as the expression `r._value  + 1` implies that `r._value` is an integer. This error must be an issue with the Go compiler for Flux that gets confused with the different types variables on the record.

I also added a similar query as an acceptance test case and am able to reproduce the panic:


```
package universe_test


import "testing"
import "csv"

testcase map_panic {
    inData = "
#datatype,string,long,dateTime:RFC3339,double,string,string,string
#group,false,false,false,false,true,true,true
#default,_result,,,,,,
,result,table,_time,_value,_field,_measurement,host
,,0,2018-05-22T19:53:26Z,10.0,load1,system,host.local
,,0,2018-05-22T19:53:36Z,10.1,load1,system,host.local
,,0,2018-05-22T19:53:46Z,10.2,load1,system,host.local
"
    outData = "
#datatype,string,long,double
#group,false,false,false
#default,_result,,
,result,table,newValue
,,0,11.0
,,0,11.1
,,0,11.2
"

    want = csv.from(csv: outData)
    got = csv.from(csv: inData)
        |> map(
            fn: (r) => {
                a = r.a
                b = r.b
                c = r._value + 1

                return {r with _value: c}
            },
        )

    testing.diff(got: got, want: want)
}

```

**Comment by nathanielc** [comment]
So turns out the bug is actually on the Rust side of things. The Rust type inference pass does not correctly infer the type of `r` as the input parameter. In the case above the `_value` property of `r` is not inferred to exist. Therefore when Go checks the input record against the input parameter type is doesn't see a type conflict for `_value` since its not in the input type. 

The bug in Rust is that for some reason no more than 2 member expressions get associated with the record. This test cases reproduces the problem:


```
    test_infer! {
        env: map![
            "r" => "{A with a: int}",
        ],
        src: r#"
            a = r.a
            b = r.b + 1
            c = r.c * 1.0
            x = r
        "#,
        exp: map![
            "a" => "int",
            "b" => "int",
            "c" => "float",
            "x" => "{A with a: int, b: int, c: float}",
        ],
    }
```

basically we would expect `x` to have properties `a`,`b` and `c` but it only gets `a`. 

This is the area of code where I expect the problem to be https://github.com/influxdata/flux/blob/master/libflux/flux-core/src/semantic/types.rs#L939-L949

That code is responsible for noticing that two records are supposed to be equal but we are comparing mismatched labels, therefore we need to apply a transitive rule and unify the type var extension of one record contains the other record. Aka rule 5 here https://github.com/influxdata/flux/blob/master/libflux/flux-core/src/semantic/types.rs#L871
